client/db/src/storage_cache.rs

```diff
@@ -668,6 +668,213 @@ impl<S: StateBackend<HasherFor<B>>, B: BlockT> StateBackend<HasherFor<B>> for Ca
 	}
 }
 
+/// Extended [`CachingState`] that will sync the caches on drop.
+pub struct SyncingCachingState<S, Block: BlockT> {
+	/// The usage statistics of the backend. These will be updated on drop.
+	state_usage: Arc<StateUsageStats>,
+	/// Reference to the meta db.
+	meta: Arc<RwLock<Meta<NumberFor<Block>, Block::Hash>>>,
+	/// Mutex to lock get exlusive access to the backend.
+	lock: Arc<RwLock<()>>,
+	/// The wrapped caching state.
+	///
+	/// This is required to be a `Option`, because sometimes we want to extract
+	/// the cache changes and Rust does not allow to move fields from types that
+	/// implement `Drop`.
+	caching_state: Option<CachingState<S, Block>>,
```

Comment:
Instead you could do something like 
```rust 
impl CachigState {
  fn drain(&mut self) -> LocalCache {
    mem::replace(&mut *self.local_cache.write(), Default::default())
  }
}

Reply:
We use `sync_cache` inside of `try_commit_operation`. So, this will not work.